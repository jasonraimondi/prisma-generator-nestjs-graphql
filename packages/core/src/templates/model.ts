import { AUTO_GENERATED_MESSAGE, ModelOptions, MyDMMF, toPrismaDocBloc } from "../constants";
import { formatWithPrettier } from "../utils/writeFile";

export async function generateModelTemplate(model: MyDMMF, config: ModelOptions) {
  return await formatWithPrettier(`
${AUTO_GENERATED_MESSAGE}
${model.imports.uuid ? "import { v4 as uuid } from 'uuid';" : ""}
${model.imports.cuid ? "import cuid from 'cuid';" : ""}
${model.imports.graphqlJSONImport ? "import GraphQLJSON from 'graphql-type-json';" : ""}
${model.imports.classValidator ? `import { ${model.imports.classValidator} } from "class-validator";` : ""}
import {
  ObjectType,
  Field,
  ${model.imports.id ? "ID," : ""}
  ${model.imports.hideField ? "HideField," : ""}
  ${model.imports.nestGraphql ? model.imports.nestGraphql : ""}
} from "@nestjs/graphql";
import {
  ${model.imports.graphqlJSONImport ? "Prisma," : ""}
  ${model.name} as Prisma${model.name},
  ${model.imports.enums}
} from "${config.clientPath}";
${model.imports.relations
  .map(
    r => `import { ${config.modelPrefix + r.type + config.modelSuffix} } from "./${r.type}${config.modelFileSuffix}";`,
  )
  .join("\n")}

export type ${model.name}Constructor = {
  ${model.fields.map(f => `${f.name}${f.isOptional ? "?" : ""}: ${f.type}${f.isOptional ? " | null" : ""}`).join("\n")}
}

@ObjectType(${config.useAbstractModels ? "{ isAbstract: true }" : ""})
export class ${model.fullName} implements Prisma${model.name} {
  ${model.fields
    .map(
      f => `
     ${f.validationBlocks ? f.validationBlocks : ""}
     ${f.isHidden ? "@HideField()" : f.graphqlField}
     ${f.isReadOnly ? "readonly " : ""}${f.name}${f.isRequired ? ":" : ": null |"} ${f.type};
  `,
    )
    .join("\n")}

  constructor(model: ${model.name}Constructor) {
    ${model.fields.map(f => `this.${f.name} = model.${f.name}${f.defaultValue ? f.defaultValue : ""};`).join("\n")}
  }
 
  static fromPrisma(data: Prisma${model.name}): ${model.fullName} {
    return new ${model.fullName}(data);
  }
  
  ${toPrismaDocBloc(model.name)}
  toPrisma(): Prisma${model.name} {
    const { ${model.imports.relations.length ? model.imports.relations.map(r => r.name) + "," : ""} ...entity } = this;
    return entity;
}
}

export { Prisma${model.name} };
  `);
}
