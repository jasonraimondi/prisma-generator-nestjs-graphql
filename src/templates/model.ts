import { transformDMMF } from "../utils/transformDMMF";
import { AUTO_GENERATED_MESSAGE, ModelOptions } from "../constants";
import { formatWithPrettier } from "../utils/writeFile";

export async function generateModelTemplate(model: ReturnType<typeof transformDMMF>[number], config: ModelOptions) {
  const { clientPath = "@prisma/client", prefix = "", abstract = true } = config;

  const template = `
    ${AUTO_GENERATED_MESSAGE}
    ${model.imports.uuid ? "import { v4 as uuid } from 'uuid';" : ""}
    ${model.imports.cuid ? "import cuid from 'cuid';" : ""}
    ${model.imports.graphqlJSONImport ? "import GraphQLJSON from 'graphql-type-json';" : ""}
    ${model.imports.classValidator ? `import { ${model.imports.classValidator} } from "class-validator";` : ""}
    import {
      ObjectType,
      Field,
      ${model.imports.id ? "ID," : ""}
      ${model.imports.hideField ? "HideField," : ""}
      ${model.imports.nestGraphql ? model.imports.nestGraphql : ""}
    } from "@nestjs/graphql";
    import {
      ${model.imports.graphqlJSONImport ? "Prisma," : ""}
      ${model.name} as Prisma${model.name},
      ${model.imports.enums}
    } from "${clientPath}";
    ${model.imports.relations.map(r => `import { ${prefix + r.type} } from "./${r.type}.model";`).join("\n")}
    
    export type ${model.name}Constructor = {
      ${model.fields
        .map(f => `${f.name}${f.isOptional ? "?" : ""}: ${f.type}${f.isOptional ? " | null" : ""}`)
        .join("\n")}
    }
    
    @ObjectType(${abstract ? "{ isAbstract: true }" : ""})
    export class ${prefix}${model.name} implements Prisma${model.name} {
      ${model.fields
        .map(
          f => `
         ${f.isHidden ? "@HideField()" : f.graphqlField}
         ${f.isReadOnly ? "readonly " : ""}${f.name}${f.isRequired ? ":" : ": null |"} ${f.type};
      `,
        )
        .join("\n")}
    
      constructor(model: ${model.name}Constructor) {
        ${model.fields.map(f => `this.${f.name} = model.${f.name}${f.defaultValue ? f.defaultValue : ""};`).join("\n")}
      }
     
      static fromPrisma(data: Prisma${model.name}): ${prefix}${model.name} {
        return new ${prefix}${model.name}(data);
      }
      
      // this method removes all relational fields from the entity, and returns the base PrismaModel
      toPrisma(): Prisma${model.name} {
        const { ${
          model.imports.relations.length ? model.imports.relations.map(r => r.name) + "," : ""
        } ...entity } = this;
        return entity;
    }
  }
   
  export { Prisma${model.name} };
  `;

  return await formatWithPrettier(template);
}
