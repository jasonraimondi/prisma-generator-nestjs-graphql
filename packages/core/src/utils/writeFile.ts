import fs from "node:fs/promises";
import path from "node:path";

import prettier from "prettier";
import { transpileModule, ModuleKind } from "typescript";

export async function writeFile(writeLocation: string, content: any, compileJs: boolean) {
  await fs.mkdir(path.dirname(writeLocation), {
    recursive: true,
  });

  if (compileJs && writeLocation.endsWith(".ts")) {
    let jsContent = transpileTypescript(content);
    jsContent = await formatWithPrettier(jsContent);
    const jsWriteLocation = writeLocation.replace(".ts", ".js");
    await fs.writeFile(jsWriteLocation, jsContent);
  }

  content = await formatWithPrettier(content);
  await fs.writeFile(writeLocation, content);
}

export function transpileTypescript(source: string) {
  const transpileOptions = { compilerOptions: { module: ModuleKind.CommonJS } };
  return transpileModule(source, transpileOptions).outputText;
}

export async function formatWithPrettier(content: string): Promise<string> {
  try {
    const prettierRc = path.join(__dirname, "../../");
    const options = await prettier.resolveConfig(prettierRc).catch();
    return await prettier.format(content, {
      ...options,
      parser: "typescript",
    });
  } catch (error) {
    return content;
  }
}
