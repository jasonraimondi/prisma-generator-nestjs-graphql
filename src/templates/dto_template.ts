import { DMMF } from "@prisma/generator-helper";
import {
  importRelations,
  graphqlType,
  shouldHide,
  isRequired,
  needsIDField,
  type,
  validationBlocks,
  importValidations,
} from "./model_template";

// @ts-ignore
export const isAutoIncrement = (field: DMMF.Field): boolean => field.default?.name === "autoincrement";

export const autoGeneratedMessage = `
/////////////////////////////////////
// DO NOT EDIT THIS FILE DIRECTLY,
// THIS FILE WAS AUTO GENERATED.
/////////////////////////////////////


`;

type GenerateDtoTemplateArgs = {
  clientPath: string;
  prefix: string;
}

export function generateDtoTemplate(args: GenerateDtoTemplateArgs, model: DMMF.Model) {
  const { clientPath, prefix } = args;

  const createField = (field: DMMF.Field) => {
    const skip = field.isUpdatedAt || field.name === "createdAt";

    if (skip) return "";

    if (field.isId) {
      const required = field.isRequired && !field.hasDefaultValue;
      // @ts-ignore
      return `
          ${validationBlocks(field.documentation)}
          @Field(() => ID, { nullable: ${!required} })  
          ${field.name}${required ? "!" : "?"}: ${type(field, { prefix })}
        `;
    }

    const required = field.name !== "createdAt" && isRequired(field);

    return `
      ${validationBlocks(field.documentation)}
      @Field(() => ${graphqlType(field, { prefix })}, { nullable: ${!required} })  
      ${field.name}${required ? "!" : "?"}: ${type(field, { prefix })}
    `;
  };

  const updateField = (field: DMMF.Field) => {
    if (shouldHide(field.documentation) || field.isUpdatedAt || field.name.startsWith("created")) return "";
    return `
@Field(() => ${field.isId ? "ID!" : graphqlType(field, {
      forceOptional: true,
      prefix,
    })}, { nullable: ${!field.isId} })  
${field.name}${field.isId ? "!" : "?"}: ${type(field, { prefix })}
`;
  };

  const whereField = (field: DMMF.Field) => {
    return `
@Field(() => ${field.isId ? "ID" : graphqlType(field, {
      forceOptional: true,
      prefix,
    })}, { nullable: true })  
${field.name}?: ${type(field, { prefix })}
`;
  };

  const hasRelatedFields = (f: DMMF.Field) =>
    !model.fields
      .filter(f => f.relationName)
      .map(f => f.name)
      .filter(relation => f.name.startsWith(relation)).length;

  const addEnumImports = (clientPath: string, model: DMMF.Model) => {
    const enums = model.fields.filter(f => f.kind == "enum");
    return enums.length ? `import { ${enums.map(f => f.type)} } from "${clientPath}";` : "";
  };

  const extraInputs = (model: DMMF.Model) => {
    const inputs: string[] = [];
    model.fields
      .filter(f => !f.isId)
      .filter(f => !f.relationName)
      .filter(f => !f.isUpdatedAt)
      .forEach(f => {
        if (f.type === "Int") inputs.push("Int");
      });
    return [...new Set(inputs)];
  };

  return `
${autoGeneratedMessage}
  
import { 
  Field, 
  InputType, 
  ${needsIDField(model) ? "ID," : ""} 
  ${extraInputs(model)} 
} from "@nestjs/graphql";
${addEnumImports(clientPath, model)}
${importRelations(model, { filterOutRelations: true, prefix })}
${importValidations(model)}
import { ${model.name}Constructor } from "./${model.name}.model";
import { PaginatorInputs } from "./paginator";

@InputType()
export class ${model.name}CreateInput implements ${model.name}Constructor {
${model.fields
    .filter(f => !f.relationName)
    .filter(f => !f.isUpdatedAt)
    .map(createField)
    .join("")}
}

@InputType()
export class ${model.name}UpdateInput {
${model.fields
    .filter(f => !f.relationName)
    .filter(f => !f.isUpdatedAt)
    .filter(hasRelatedFields)
    .map(updateField)
    .join("")}
}

@InputType()
export class ${model.name}WhereInput {
${model.fields
    .filter(f => !f.relationName)
    .filter(f => !f.isUpdatedAt)
    .filter(f => !f.name.startsWith("created"))
    .filter(f => !shouldHide(f.documentation))
    .map(whereField)
    .join("")}
}

@InputType()
export class ${model.name}PaginatorInput extends PaginatorInputs {
}
`;
}
