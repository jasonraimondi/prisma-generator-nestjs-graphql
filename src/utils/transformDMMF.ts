import { DMMF } from "@prisma/generator-helper";

import { getDefaultValue, graphqlField, graphqlType, type } from "./formatter";
import { ModelOptions } from "../constants";

export function checkIsID(f: DMMF.Field): boolean {
  return f.isId;
}

export function checkTypeJson(f: DMMF.Field): boolean {
  return f.type === "Json";
}

export function checkCUID(f: DMMF.Field): boolean {
  return typeof f.default === "object" && "name" in f.default && f.default.name === "cuid";
}

export function checkUUID(f: DMMF.Field): boolean {
  return typeof f.default === "object" && "name" in f.default && f.default.name === "uuid";
}

export function checkAutoIncrement(f: DMMF.Field): boolean {
  return typeof f.default === "object" && "name" in f.default && f.default.name === "autoincrement";
}

export function checkRequired(f: DMMF.Field): boolean {
  return (f.isRequired || f.isId) && !f.relationName;
}

export function checkReadOnly(f: DMMF.Field): boolean {
  return f.isReadOnly || f.isId;
}

export function checkOptional(f: DMMF.Field): boolean {
  return f.isList || !f.isRequired || f.hasDefaultValue || Boolean(f.relationName);
}

export function checkHidden(f: DMMF.Field): boolean {
  return f.documentation?.includes("@HideField()") ?? false;
}

export function checkCanCreate(f: DMMF.Field): boolean {
  return !(f.relationName || f.isUpdatedAt);
}

export function checkCanUpdate(f: DMMF.Field): boolean {
  return !(checkHidden(f) || f.isReadOnly || f.isId || f.relationName || f.isUpdatedAt || f.name.startsWith("created"));
}

export function extractValidations(model: DMMF.Model) {
  let validations: string[] = [];
  const regex = /@Validate.[a-zA-Z]+/g;
  model.fields
    .map(f => f.documentation?.match(regex) ?? [])
    .forEach(s => s.forEach(t => validations.push(t.replace("@Validate.", ""))));
  validations = [...new Set(validations)];
  return validations.length ? validations : false;
}

export function extractNestGraphql(model: DMMF.Model) {
  let nestGraphqlImports: string[] = [];
  model.fields
    // .filter(f => !f.isId)
    .filter(f => !f.relationName)
    .filter(f => !f.isUpdatedAt)
    .forEach(f => {
      if (f.type === "Int") nestGraphqlImports.push("Int");
    });
  nestGraphqlImports = [...new Set(nestGraphqlImports)];
  return nestGraphqlImports.length ? nestGraphqlImports : false;
}

export function transformDMMF(dmmf: DMMF.Document, options: ModelOptions) {
  return dmmf.datamodel.models.map(model => {
    return {
      name: model.name,
      fullName: options.modelPrefix + model.name + options.modelSuffix,
      imports: {
        id: model.fields.some(checkIsID),
        cuid: model.fields.some(checkCUID),
        uuid: model.fields.some(checkUUID),
        graphqlJSONImport: model.fields.some(checkTypeJson),
        hideField: model.fields.some(checkHidden),
        nestGraphql: extractNestGraphql(model),
        classValidator: extractValidations(model),
        enums: model.fields.filter(f => f.kind == "enum").map(f => f.type),
        relations: model.fields
          .filter(f => f.relationName)
          .map(f => ({
            name: f.name,
            type: f.type,
          })),
      },
      fields: model.fields.map(f => {
        return {
          name: f.name,
          kind: f.kind,
          type: type(f, options),
          graphqlType: graphqlType(f, options),
          graphqlField: graphqlField(f, options),
          defaultValue: getDefaultValue(f),
          isId: f.isId,
          isHidden: checkHidden(f),
          isRequired: checkRequired(f),
          isReadOnly: checkReadOnly(f),
          isAutoIncrement: checkAutoIncrement(f),
          isUUID: checkUUID(f),
          isCUID: checkCUID(f),
          isOptional: checkOptional(f),
          isRelatedField: Boolean(f.relationName),
          canCreate: checkCanCreate(f),
          canUpdate: checkCanUpdate(f),

          validationBlocks: f.documentation
            // @see https://regex101.com/r/7Y548N/1
            ?.match(/@Validate.[a-zA-Z]+\([a-zA-Z ,()'"[\]]+/g)
            ?.map(m => m.split("@Validate.")[1])
            .map(m => `@${m}`)
            .join("\n") as string | undefined,
        };
      }),
    };
  });
}
