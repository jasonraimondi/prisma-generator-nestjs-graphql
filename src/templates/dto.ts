import { DMMF } from "@prisma/generator-helper";
import { AUTO_GENERATED_MESSAGE, ModelOptions } from "../constants";
import { transformDMMF } from "../utils/transformDMMF";
import {
  graphqlType,
  importRelations,
  isRequired,
  needsGraphqlJSONImport,
  shouldHide,
  type,
  validationBlocks,
} from "../utils/formatter";
import { formatWithPrettier } from "../utils/writeFile";

type GenerateDtoTemplateArgs = {
  clientPath: string;
  prefix: string;
};

export function xgenerateDtoTemplate(args: GenerateDtoTemplateArgs, model: DMMF.Model) {
  const { clientPath, prefix } = args;

  const createField = (field: DMMF.Field) => {
    const skip = field.isUpdatedAt || field.name === "createdAt";

    if (skip) return "";

    if (field.isId) {
      const required = field.isRequired && !field.hasDefaultValue;
      // @ts-ignore
      return `
          ${validationBlocks(field.documentation)}
          @Field(() => ID, { nullable: ${!required} })  
          ${field.name}${required ? "!" : "?"}: ${type(field, { prefix })}
        `;
    }

    const required = field.name !== "createdAt" && isRequired(field);

    return `
      ${validationBlocks(field.documentation)}
      @Field(() => ${graphqlType(field, { prefix })}, { nullable: ${!required} })  
      ${field.name}${required ? "!" : "?"}: ${type(field, { prefix })}
    `;
  };

  const updateField = (field: DMMF.Field) => {
    if (shouldHide(field.documentation) || field.isUpdatedAt || field.name.startsWith("created")) return "";
    return `
@Field(() => ${
      field.isId
        ? "ID!"
        : graphqlType(field, {
            forceOptional: true,
            prefix,
          })
    }, { nullable: ${!field.isId} })  
${field.name}${field.isId ? "!" : "?"}: ${type(field, { prefix })}
`;
  };

  const whereField = (field: DMMF.Field) => {
    return `
@Field(() => ${
      field.isId
        ? "ID"
        : graphqlType(field, {
            forceOptional: true,
            prefix,
          })
    }, { nullable: true })  
${field.name}?: ${type(field, { prefix })}
`;
  };

  const listEnums = model.fields.filter(f => f.kind == "enum");

  const needsPrismaImport = (model: DMMF.Model) => {
    if (needsGraphqlJSONImport(model)) return true;
    if (listEnums.length > 0) return true;
  };

  return `
    ${
      needsPrismaImport(model)
        ? `import {
      ${needsGraphqlJSONImport(model) ? "Prisma," : ""}
      ${listEnums.map(f => f.type)}
    } from "${clientPath}";`
        : ""
    }
    ${importRelations(model, { filterOutRelations: true, prefix })}
    
    @InputType()
    export class ${model.name}CreateInput implements ${model.name}Constructor {
    ${model.fields
      .filter(f => !f.relationName)
      .filter(f => !f.isUpdatedAt)
      .map(createField)
      .join("")}
    }
    
    @InputType()
    export class ${model.name}UpdateInput {
    ${model.fields
      .filter(f => !f.relationName)
      .filter(f => !f.isUpdatedAt)
      .map(updateField)
      .join("")}
    }
    
    @InputType()
    export class ${model.name}WhereInput {
    ${model.fields
      .filter(f => !f.relationName)
      .filter(f => !f.isUpdatedAt)
      .filter(f => !f.name.startsWith("created"))
      .filter(f => !shouldHide(f.documentation))
      .map(whereField)
      .join("")}
    }
    
    @InputType()
    export class ${model.name}PaginatorInput extends PaginatorInputs {
    }
`;
}

export async function generateDtoTemplate(m: ReturnType<typeof transformDMMF>[number], config: ModelOptions) {
  const { clientPath = "@prisma/client", prefix = "" } = config;

  const template = `
    ${AUTO_GENERATED_MESSAGE}
    ${m.imports.graphqlJSONImport ? "import GraphQLJSON from 'graphql-type-json';" : ""}
    ${m.imports.classValidator ? `import { ${m.imports.classValidator} } from "class-validator";` : ""}
    import { 
      Field, 
      InputType, 
      ${m.imports.id ? "ID," : ""} 
      ${m.imports.nestGraphql ? m.imports.nestGraphql : ""} 
    } from "@nestjs/graphql";
    ${m.imports.relations.map(r => `import { ${config.prefix + r.type} } from "./${r.type}.model";`).join("\n")}
    import {
      ${m.imports.graphqlJSONImport ? "Prisma," : ""}
      ${m.name} as Prisma${m.name},
      ${m.imports.enums}
    } from "${clientPath}";
    import { PaginatorInputs } from "./paginator";
    import { ${m.name}Constructor } from "./${m.name}.model";
    
    @InputType()
    export class ${m.name}CreateInput implements ${m.name}Constructor {
    ${m.fields
      .filter(f => f.canCreate)
      .map(
        f => `
          ${f.validationBlocks ? f.validationBlocks : ""}
          @Field(() => ${f.isId ? "ID" : f.graphqlType}, { nullable: ${f.isId || !f.isRequired} })  
          ${f.name}${f.isId || !f.isRequired ? "?" : "!"}: ${f.type};
        `,
      )
      .join("\n")}
    }
    
    @InputType()
    export class ${m.name}UpdateInput {
      @Field(() => ID, { nullable: true })  
      ${m.fields.find(f => f.isId)!.name}?: ${m.fields.find(f => f.isId)!.type}
    
    ${m.fields
      .filter(f => f.canUpdate)
      .map(
        f => `
          @Field(() => ${f.graphqlType}, { nullable: ${f.isOptional} })
          ${f.name}${f.isId ? "!" : "?"}: ${f.type}
        `,
      )
      .join("\n")}
    }

    @InputType()
    export class ${m.name}WhereInput {
    ${m.fields
      .filter(f => !f.isRelatedField)
      .filter(f => !f.name.startsWith("created"))
      .filter(f => !f.isHidden)
      .map(
        f => `
          @Field(() => ${f.isId ? "ID" : f.graphqlType}, { nullable: true })  
          ${f.name}?: ${f.type}
        `,
      )
      .join("\n")}
    }

    @InputType()
      export class ${m.name}PaginatorInput extends PaginatorInputs {
    }
  `;

  return await formatWithPrettier(template);
}
