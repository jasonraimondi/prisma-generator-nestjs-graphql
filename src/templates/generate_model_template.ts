import { transformDMMF } from "../utils/transformDMMF";
import { AUTO_GENERATED_MESSAGE } from "../constants/constants";

export type ModelOptions = {
  clientPath?: string;
  prefix?: string;
  abstract?: boolean;
  modelSuffix?: string;
};

export function generateModelTemplate(m: ReturnType<typeof transformDMMF>[number], config: ModelOptions = {}) {
  const { clientPath = "@prisma/client", modelSuffix = ".model", prefix = "", abstract = true } = config;

  return `
    ${AUTO_GENERATED_MESSAGE}
    ${m.imports.uuid ? "import { v4 as uuid } from 'uuid';" : ""}
    ${m.imports.cuid ? "import cuid from 'cuid';" : ""}
    ${m.imports.graphqlJSONImport ? "import GraphQLJSON from 'graphql-type-json';" : ""}
    ${m.imports.classValidator ? `import { ${m.imports.classValidator} } from "class-validator";` : ""}
    import {
      ObjectType,
      Field,
      ${m.imports.id ? "ID," : ""}
      ${m.imports.hideField ? "HideField," : ""}
      ${m.imports.nestGraphql ? m.imports.nestGraphql : ""}
    } from "@nestjs/graphql";
    import {
      ${m.imports.graphqlJSONImport ? "Prisma," : ""}
      ${m.name} as Prisma${m.name},
      ${m.imports.enums}
    } from "${clientPath}";
    ${m.imports.relations.map(r => `import { ${r.type} } from "./${r.type}${modelSuffix}";`).join("\n")}
    
    export type ${m.name}Constructor = {
      ${m.fields.map(f => `${f.name}${f.isOptional ? "?" : ""}: ${f.type}${f.isOptional ? " | null" : ""}`).join("\n")}
    }
    
    @ObjectType(${abstract ? "{ isAbstract: true }" : ""})
    export class ${prefix}${m.name} implements Prisma${m.name} {
    
      ${m.fields
        .map(
          f => `
         ${f.isHidden ? "@HideField()" : f.graphqlFields}
         ${f.isReadOnly ? "readonly " : ""}${f.name}${f.isRequired ? ":" : ": null |"} ${f.type};
      `,
        )
        .join("\n")}
    
      constructor(model: ${m.name}Constructor) {
        ${m.fields.map(f => `this.${f.name} = model.${f.name}${f.defaultValue};`).join("\n")}
      }
     
      static fromPrisma(data: Prisma${m.name}): ${prefix}${m.name} {
        return new ${prefix}${m.name}(data);
      }
      
      // this method removes all relational fields from the entity, and returns the base PrismaModel
      toPrisma(): Prisma${m.name} {
        const { ${m.imports.relations.length ? m.imports.relations.map(r => r.name) + "," : ""} ...entity } = this;
        return entity;
    }
  }
   
  export { Prisma${m.name} };
  `;
}
