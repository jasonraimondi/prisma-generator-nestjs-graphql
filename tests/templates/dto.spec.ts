import { describe, it, expect } from "vitest";
import { getDMMF } from "@prisma/internals";

import { transformDMMF } from "../../src/utils/transformDMMF";
import { generateDtoTemplate } from "../../src/templates/dto";

describe("dto template", () => {
  const config = { prefix: "Base", abstract: true, clientPath: "./client-path" };

  describe("with relation", () => {
    it("success", async () => {
      const dmmf = await getDMMF({
        datamodel: `
          model User {
            id             String @id @default(uuid())
            favoriteNumber Int
            posts          Post[]
          }
          model Post {
            id       String @id @default(uuid())
            userId   String
            user     User   @relation(fields: [userId], references: [id])
          }
        `,
      });
      const [userModel] = transformDMMF(dmmf, config);
      const template = await generateDtoTemplate(userModel, config);

      console.log(userModel, template);

      // expect(template).toMatchSnapshot();
    });
  });

  describe.skip("uuid", () => {
    it("success", async () => {
      const dmmf = await getDMMF({
        datamodel: `
          model UuidExample {
            id String @id @default(uuid())
          }
        `,
      });

      const [userModel] = transformDMMF(dmmf, config);
      const modelTemplate = await generateDtoTemplate(userModel, config);

      // expect(modelTemplate).toMatchSnapshot();
    });
  });
});
