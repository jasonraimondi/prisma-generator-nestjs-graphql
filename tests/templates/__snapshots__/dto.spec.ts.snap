// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dto templater > with autoincrement 1`] = `
"/////////////////////////////////////
// THIS FILE WAS AUTO GENERATED
// DO NOT EDIT THIS FILE DIRECTLY
/////////////////////////////////////

import { Field, InputType, ID, Int } from \\"@nestjs/graphql\\";
import {} from \\"./client-path\\";
import { PaginatorInputs } from \\"../paginator\\";
import { UserConstructor } from \\"../models/User\\";

@InputType()
export class UserCreateInput implements UserConstructor {
  @Field(() => ID, { nullable: false })
  id!: number;
}

@InputType()
export class UserUpdateInput {
  @Field(() => ID, { nullable: true })
  id?: number;
}

@InputType()
export class UserWhereInput {
  @Field(() => ID, { nullable: true })
  id?: number;
}

@InputType()
export class UserPaginatorInput extends PaginatorInputs {}
"
`;

exports[`dto templater > with cuid 1`] = `
"/////////////////////////////////////
// THIS FILE WAS AUTO GENERATED
// DO NOT EDIT THIS FILE DIRECTLY
/////////////////////////////////////

import { Field, InputType, ID } from \\"@nestjs/graphql\\";
import {} from \\"./client-path\\";
import { PaginatorInputs } from \\"../paginator\\";
import { UserConstructor } from \\"../models/User\\";

@InputType()
export class UserCreateInput implements UserConstructor {
  @Field(() => ID, { nullable: false })
  id!: string;
}

@InputType()
export class UserUpdateInput {
  @Field(() => ID, { nullable: true })
  id?: string;
}

@InputType()
export class UserWhereInput {
  @Field(() => ID, { nullable: true })
  id?: string;
}

@InputType()
export class UserPaginatorInput extends PaginatorInputs {}
"
`;

exports[`dto templater > with integer 1`] = `
"/////////////////////////////////////
// THIS FILE WAS AUTO GENERATED
// DO NOT EDIT THIS FILE DIRECTLY
/////////////////////////////////////

import { Field, InputType, ID, Int } from \\"@nestjs/graphql\\";
import {} from \\"./client-path\\";
import { PaginatorInputs } from \\"../paginator\\";
import { UserConstructor } from \\"../models/User\\";

@InputType()
export class UserCreateInput implements UserConstructor {
  @Field(() => ID, { nullable: false })
  id!: string;

  @Field(() => Int!, { nullable: false })
  favoriteNumber!: number;
}

@InputType()
export class UserUpdateInput {
  @Field(() => ID, { nullable: true })
  id?: string;

  @Field(() => Int!, { nullable: false })
  favoriteNumber?: number;
}

@InputType()
export class UserWhereInput {
  @Field(() => ID, { nullable: true })
  id?: string;

  @Field(() => Int!, { nullable: true })
  favoriteNumber?: number;
}

@InputType()
export class UserPaginatorInput extends PaginatorInputs {}
"
`;

exports[`dto templater > with json 1`] = `
"/////////////////////////////////////
// THIS FILE WAS AUTO GENERATED
// DO NOT EDIT THIS FILE DIRECTLY
/////////////////////////////////////

import GraphQLJSON from \\"graphql-type-json\\";

import { Field, InputType, ID, Int } from \\"@nestjs/graphql\\";
import { Prisma } from \\"./client-path\\";
import { PaginatorInputs } from \\"../paginator\\";
import { UserConstructor } from \\"../models/User\\";

@InputType()
export class UserCreateInput implements UserConstructor {
  @Field(() => ID, { nullable: false })
  id!: number;

  @Field(() => GraphQLJSON!, { nullable: false })
  jsonCol!: Prisma.JsonValue;
}

@InputType()
export class UserUpdateInput {
  @Field(() => ID, { nullable: true })
  id?: number;

  @Field(() => GraphQLJSON!, { nullable: false })
  jsonCol?: Prisma.JsonValue;
}

@InputType()
export class UserWhereInput {
  @Field(() => ID, { nullable: true })
  id?: number;

  @Field(() => GraphQLJSON!, { nullable: true })
  jsonCol?: Prisma.JsonValue;
}

@InputType()
export class UserPaginatorInput extends PaginatorInputs {}
"
`;

exports[`dto templater > with relation 1`] = `
"/////////////////////////////////////
// THIS FILE WAS AUTO GENERATED
// DO NOT EDIT THIS FILE DIRECTLY
/////////////////////////////////////

import { Field, InputType, ID } from \\"@nestjs/graphql\\";
import {} from \\"./client-path\\";
import { PaginatorInputs } from \\"../paginator\\";
import { UserConstructor } from \\"../models/User\\";

@InputType()
export class UserCreateInput implements UserConstructor {
  @Field(() => ID, { nullable: false })
  id!: string;
}

@InputType()
export class UserUpdateInput {
  @Field(() => ID, { nullable: true })
  id?: string;
}

@InputType()
export class UserWhereInput {
  @Field(() => ID, { nullable: true })
  id?: string;
}

@InputType()
export class UserPaginatorInput extends PaginatorInputs {}
"
`;

exports[`dto templater > with uuid 1`] = `
"/////////////////////////////////////
// THIS FILE WAS AUTO GENERATED
// DO NOT EDIT THIS FILE DIRECTLY
/////////////////////////////////////

import { Field, InputType, ID } from \\"@nestjs/graphql\\";
import {} from \\"./client-path\\";
import { PaginatorInputs } from \\"../paginator\\";
import { UserConstructor } from \\"../models/User\\";

@InputType()
export class UserCreateInput implements UserConstructor {
  @Field(() => ID, { nullable: false })
  id!: string;
}

@InputType()
export class UserUpdateInput {
  @Field(() => ID, { nullable: true })
  id?: string;
}

@InputType()
export class UserWhereInput {
  @Field(() => ID, { nullable: true })
  id?: string;
}

@InputType()
export class UserPaginatorInput extends PaginatorInputs {}
"
`;
